Library/Homebrew/cask/cask_loader.rb

```diff
@@ -589,6 +602,32 @@ def self.for(ref, need_path: false, warn: true)
       end
     end
 
+    sig { params(ref: String, config: T.nilable(Config), warn: T::Boolean).returns(Cask) }
+    def self.load_installed_cask(ref, config: nil, warn: true)
+      tap, token = Tap.with_cask_token(ref)
+      token ||= ref
+      tap ||= Cask.new(ref).tab.tap
+
+      if tap.nil?
+        self.load(token, config:, warn:)
+      else
+        begin
+          self.load("#{tap}/#{token}", config:, warn:)
+        rescue CaskUnavailableError
+          # cask may be migrated to different tap. Try to search in all taps.
+          self.load(token, config:, warn:)
+        end
+      end
+    end
```

Comment:
Discussion: Do we plan on using this outside of `Cask::Caskroom.cask`? If not, the logic in https://github.com/Homebrew/brew/pull/17821 seems simpler because it doesn't have to worry about every possible cask reference type as input.

I can see that being able to specify that you only want to load installed casks would be useful for different commands though and I'm not sure how to elegantly handle that (one of the reasons my PR is still marked as draft).

Reply:
One assumption I made is that we can use the full name here only if the tap is installed. If it's not installed, we just default to the short token because the original tap has the migration data too so using the full name won't be helpful in that case either.

Reply:
I also think it's confusing that it could potentially return a cask that is not installed.

Reply:
> Do we plan on using this outside of `Cask::Caskroom.cask`?

Hard to predict. `Formulary::from_keg` (the formula analog) is used in the linkage checked and `brew outdated`. Both of those use cases happen to be formula-specific, but I think being able to default prefer loading an installed cask is something that other places could find useful.

> I also think it's confusing that it could potentially return a cask that is not installed.

That's fair. How about changing the name to `CaskLoader::load_prefer_installed`?

I'll make that change